File: backend/src/main/java/com/headless/ecommerce/service/ProductService.java
package com.headless.ecommerce.service;

import com.headless.ecommerce.dto.request.ProductCreateRequest;
import com.headless.ecommerce.dto.request.ProductUpdateRequest;
import com.headless.ecommerce.dto.response.PageResponse;
import com.headless.ecommerce.dto.response.ProductResponse;
import com.headless.ecommerce.model.Product;

/**
 * Product service interface for product management.
 */
public interface ProductService {

    /**
     * Gets a paginated list of on-shelf products with optional filters.
     *
     * @param page page number (0-based)
     * @param size page size
     * @param categoryId optional category filter
     * @param keyword optional search keyword
     * @param sort sort specification
     * @return paginated product response
     */
    PageResponse<ProductResponse> getProducts(int page, int size,
                                               Long categoryId, String keyword, String sort);

    /**
     * Gets a product by ID.
     *
     * @param id the product ID
     * @return the product response
     */
    ProductResponse getProductById(Long id);

    /**
     * Creates a new product.
     *
     * @param request the product creation request
     * @return the created product response
     */
    ProductResponse createProduct(ProductCreateRequest request);

    /**
     * Updates an existing product.
     *
     * @param id the product ID
     * @param request the product update request
     * @return the updated product response
     */
    ProductResponse updateProduct(Long id, ProductUpdateRequest request);

    /**
     * Deletes a product (sets status to OFF_SHELF).
     *
     * @param id the product ID
     */
    void deleteProduct(Long id);

    /**
     * Updates a product's status.
     *
     * @param id the product ID
     * @param status the new status
     * @return the updated product response
     */
    ProductResponse updateProductStatus(Long id, String status);

    /**
     * Finds a Product entity by ID.
     *
     * @param id the product ID
     * @return the Product entity
     */
    Product findById(Long id);
}

File: frontend/src/services/product-service.ts
import api from "@/lib/api";
import { API_PATHS } from "@/lib/constants";
import type { ApiResponse, PageResponse } from "@/types/api";
import type { Product, ProductFilters, ProductCreateRequest, ProductUpdateRequest } from "@/types/product";

/**
 * Gets a paginated list of products.
 */
export async function getProducts(
  filters?: ProductFilters
): Promise<ApiResponse<PageResponse<Product>>> {
  const params: Record<string, string | number> = {};
  if (filters?.page !== undefined) params.page = filters.page;
  if (filters?.size !== undefined) params.size = filters.size;
  if (filters?.categoryId !== undefined) params.categoryId = filters.categoryId;
  if (filters?.keyword) params.keyword = filters.keyword;
  if (filters?.sort) params.sort = filters.sort;

  const response = await api.get<ApiResponse<PageResponse<Product>>>(
    API_PATHS.PRODUCTS.LIST,
    { params }
  );
  return response.data;
}

/**
 * Gets a product by ID.
 */
export async function getProductById(id: number): Promise<ApiResponse<Product>> {
  const response = await api.get<ApiResponse<Product>>(
    API_PATHS.PRODUCTS.DETAIL(id)
  );
  return response.data;
}

/**
 * Searches products by keyword.
 */
export async function searchProducts(
  keyword: string,
  page = 0,
  size = 12
): Promise<ApiResponse<PageResponse<Product>>> {
  const response = await api.get<ApiResponse<PageResponse<Product>>>(
    API_PATHS.PRODUCTS.SEARCH,
    { params: { keyword, page, size } }
  );
  return response.data;
}

/**
 * Creates a new product (admin).
 */
export async function createProduct(
  data: ProductCreateRequest
): Promise<ApiResponse<Product>> {
  const response = await api.post<ApiResponse<Product>>(
    API_PATHS.ADMIN.PRODUCTS,
    data
  );
  return response.data;
}

/**
 * Updates a product (admin).
 */
export async function updateProduct(
  id: number,
  data: ProductUpdateRequest
): Promise<ApiResponse<Product>> {
  const response = await api.put<ApiResponse<Product>>(
    API_PATHS.ADMIN.PRODUCT_DETAIL(id),
    data
  );
  return response.data;
}

/**
 * Deletes a product (admin).
 */
export async function deleteProduct(id: number): Promise<ApiResponse<void>> {
  const response = await api.delete<ApiResponse<void>>(
    API_PATHS.ADMIN.PRODUCT_DETAIL(id)
  );
  return response.data;
}

/**
 * Updates a product's status (admin).
 */
export async function updateProductStatus(
  id: number,
  status: "ON_SHELF" | "OFF_SHELF"
): Promise<ApiResponse<Product>> {
  const response = await api.put<ApiResponse<Product>>(
    API_PATHS.ADMIN.PRODUCT_STATUS(id),
    { status }
  );
  return response.data;
}
